### DISCUSSION TITLE
Experience With Tauri + SvelteKit

### DISCUSSION THREAD

**Discussion by lorenzolewis** [tauri-apps/tauri#6423] [Q&A]
Recently there's been a lot of conversation around using SvelteKit and Tauri together. We want to know:

- 💪 What's worked for you? 
- 🤕What's not quite worked as expected?
- 🤔Have there been any learning curves or tough points along the way?

Let us know your feedback!

**Comment by lorenzolewis** [comment]
_A few thoughts from looking through the steps we have in the guides: https://tauri.app/v1/guides/getting-started/setup/sveltekit_

SvelteKit and Tauri can work well together, but getting them setup can involve a few manual steps:

1. Add and setup [@sveltejs/adapter-static](https://github.com/sveltejs/kit/tree/master/packages/adapter-static)
2. Add a root `+layout.ts` file with the following to enable disable SSR and enable pre-rendering:
```ts
export const prerender = true
export const ssr = false
```

These aren't necessarily _difficult_ steps, but knowing the difference between SPA, SSG, SSR and what you can/can't do between each can make it tough to get started as a beginner

**Reply by austenc** [reply]
As someone coming from a PHP / Laravel framework background and not being overly familiar with modern js "frameworks", I admittedly followed the docs on this step without understanding what much of it meant at the time. 

The disabling of SSR makes sense (since there's no server), but those docs were confusing to me mainly because of the mention of SPAs. I wasn't sure how to know if I should dig into that part of the docs or config or not, so perhaps simply providing some examples for when you'd use each approach would clear that up.

Here are a few other thoughts / points of confusion:
- Should I use Svelte or SvelteKit / why use SvelteKit? (chose the latter because of routing and service workers)
- Is SvelteKit routing "SPA-like" or do I need to enable SPA features in the config like the Tauri docs mention?
- What areas or features of SvelteKit are off-limits or won't work in the Tauri/desktop context?

**Reply by lorenzolewis** [reply]
Thanks for giving your thoughts @austenc!

A things we could do to help would be to explain the pros/cons, features you gain/lose (in the context of Tauri) and recommended setup for:

1. Svelte vs. SvelteKit
2. SSR vs. SPA

For those who would like to know the _why_ behind those options we could do some work in the docs to explain them. Or if someone just wants to zoom through and get a Svelte app up and running they could just follow the recommendations.

**Reply by austenc** [reply]
Of course @lorenzolewis, wish I could be more helpful! Likely too much of a Tauri and SvelteKit noob to give much more input, though. 

I don't think those explanations are _mandatory_ in the docs by any means. Was definitely zooming through myself 👍 
That being said, it would be really neat to learn the pros/cons and see the differences explained by the folks who understand it well!

**Reply by jamauro** [reply]
> For those who would like to know the _why_ behind those options we could do some work in the docs to explain them.

Would love to understand more of the why behind SSG vs the other options

**Reply by sameoldlab** [reply]
I think the main reason in Tauri for SvelteKit, rather than Svelte is routing. I'm using Svelte with a lot of if blocks to render everything. SvelteKit, with SSR disabled, seems like it would be more straightforward for most people and picking a compatible routing library also requires users to know about SSR and SPAs.

Maybe something in docs that breaks down scenarios where SvelteKit becomes useful could help. Or linking to [SvelteKit docs](https://kit.svelte.dev/docs/introduction#what-does-sveltekit-provide-on-top-of-svelte) and [what you get with Svelte ](https://svelte.dev/faq#is-there-a-router) (adding nuance that links work differently outside the browser, which can be very unexpected using Tauri the first time).

**Reply by KameiKojirou** [reply]
SvelteKit is the default recommended by the Svelte Dev team. All the documentation on their end points that direction as well. I believe the small manual edits and be mostly in line with the default build is the least resistant path in my opinion. Even 'svelte@latest' currently builds SvelteKit and points at its documentation instead of regular Svelte.

**Comment by astral4** [comment]
I've made one static website with SvelteKit and figured the knowledge would transfer over to Tauri. For the most part, that was true, but there are some bumps I've encountered.

One thing that confused me at the beginning was Content Security Policy configuration. SvelteKit lets you do this via `svelte.config.js` ([SvelteKit docs](https://kit.svelte.dev/docs/configuration#csp)), but Tauri also has a CSP configuration in `tauri.conf.json` ([Tauri docs](https://tauri.app/v1/api/config#securityconfig.csp)). I ended up configuring the CSP with SvelteKit and not touching the Tauri settings. I'm also not sure how SvelteKit interacts with Tauri when it comes to injecting nonces/hashes; apparently both do this. It hasn't caused issues yet, so I haven't looked into it.

On to the config itself:
```js
// svelte.config.js

import adapter from '@sveltejs/adapter-static';
import preprocess from 'svelte-preprocess';

/** @type {import('@sveltejs/kit').Config} */
const config = {
	preprocess: preprocess(),
	kit: {
		adapter: adapter(),
		csp: {
			mode: 'hash',
			directives: {
				'default-src': ['none'],
				'img-src': ['self'],
				'script-src': ['self'],
				'style-src': ['self', 'unsafe-inline'],
			}
		}
	}
};

export default config;
```
I use a restrictive-by-default CSP. (`unsafe-inline` is there because Svelte uses inline styles for transitions.) This configuration worked fine for my previous SvelteKit website... but for my Tauri app, Vite's HMR broke! For a while, I was running `npm run tauri dev` *every time* I changed a file. Eventually, I fixed the issue by setting a `connect-src` directive like so:
```diff
        'style-src': ['self', 'unsafe-inline'],
+       // Vite uses WebSockets for HMR, so WebSocket connections to localhost:5173 are whitelisted during development
+       'connect-src': process.env.TAURI_DEBUG === 'true' ? ['ws://localhost:5173'] : undefined
```
I didn't realize that browsers and Tauri handled WebSocket connections differently.

This last point is not necessarily specific to SvelteKit and is more of a Vite thing, but I think it's worth mentioning. I encountered a bug in my app where certain images would not show in release builds, but were visible in debug builds. I think the cause goes like this: by default, Vite inlines images smaller than 4KiB as base64 during release builds ([Vite docs](https://vitejs.dev/config/build-options.html#build-assetsinlinelimit)). Then, Tauri does something with the CSP that blocks these assets.
| Mode | Browser webpage | Tauri webview |
|:-:|:-:|:-:|
| Debug | ✅ | ✅ |
| Release | ✅ | ❌ |

I resolved this issue by disabling asset inlining in the Vite config file (`vite.config.ts`) like so:
```diff
build: {
        // The Windows webview (WebView2) is supposed to update itself automatically, so it's safe to target the latest 3 versions (i.e. Blink 108).
	// The macOS webview (WebKit) only updates with new OS versions, so the build target is more conservative; the earliest Safari version for macOS Big Sur is 14 (i.e. WebKit 610.2.11).
	// The Linux webview (WebKitGTK) is based on WebKit, so the target is the same as macOS.
	// Reference: https://discord.com/channels/616186924390023171/1082055425907753063
	target: process.env.TAURI_PLATFORM == 'windows' ? 'edge108' : ['es2021', 'safari14'],
+	// By default, Vite inlines assets smaller than 4 KiB as base64 (see https://vitejs.dev/config/build-options.html#build-assetsinlinelimit).
+	// These assets are blocked by the CSP in production builds, so inlining is explicitly disabled.
+	assetsInlineLimit: 0
},
```
<br>
I would say that my experience with Tauri and SvelteKit has been positive. For the most part, it *just works*. There are just some small things to be aware of.

**Comment by acurrieclark** [comment]
If you have a single route (not single page) application, I would suggest that the most straightforward option would be to use Svelte by itself. 

If you need routing, SvelteKit seems like the obvious choice. There are 3rd party routers available (and have been for a long time) but SPA mode is supported in SvelteKit with very few configuration changes.

As it is supported by Svelte itself, it makes sense to me to go with SvelteKit in this case. That is what I have done for my current Tauri project.

**Comment by lorenzolewis** [comment]
Thanks for all of the insights! @FabianLars @probablykasper and I have been doing some research into this. Putting our notes so far below:

## Recommended Setup (SPA)

Starting out with SPA mode by using `adapter-static` and setting `index.html` as the fallback (this is important to make it go into SPA mode).

## Advanced Setup (SSG with prerendering)

Once you know your way around SvelteKit you can start enabling prerendering to improve First Contentful/Meaningful Paint metrics. This can be applied per route or globally if set in the root `+layout.js` file.

## Considerations

- Recommend not using server routes (`+server.js`, `+page.server.js`, `+layout.server.js` files)
    - More information about prerendering for server routes: https://kit.svelte.dev/docs/page-options#prerender-prerendering-server-routes
    - Always test it in `tauri build` to make sure it works at runtime
- If using `[slug]` routing be sure to read https://kit.svelte.dev/docs/page-options#prerender-when-not-to-prerender to understand the limitations
- CSP: We're still exploring this. Currently we're recommending to use either the frontend's CSP or Tauri's. We're researching into if they cause conflicts with each other. @astral4 feel free to join in and add your insight here if you'd like!

**Comment by Cobular** [comment]
I've used Sveltekit for ~10 projects and used it with Tauri for 2 or 3, I think it works pretty well! The pair is currently my favorite way to make any desktop application and I can't wait for the rest of mobile support to be ready so I can stop making PWAs for my mobile needs!

Sveltekit has first class support for ssg as others have mentioned, however it is not quite as great an experience as the other modes, especially when it comes to form submission and the like. Most benefits of SSR over SSG really do go away significantly when in an application context like this, the bundle is served off disk and so the loading time is negligable, plus the application itself routinely takes 10 or more times longer to start up than the page takes to load, so this is not the end of the world. However, the UX would still be a good 20% better if we had SSR in Tauri!

I believe the next best step forward would be a Sveltekit adapter for Tauri - one that gives Tauri first class support running the SSR code in the core process or something and serving the rendered files in such a way that developers can keep the full Sveltekit feature suite with no compromises.

Sveltekit and Tauri already share a number of goals with regard to performance and bundle size, and I'd love it if they worked together even better!

**Reply by lorenzolewis** [reply]
Thanks for the feedback, @Cobular!

Let's talk about the example you mentioned: Form Submissions. I know this is an area that SvelteKit has some awesome functionality but I have to admit I'm not well versed in it (yet). I had a quick read through of https://kit.svelte.dev/docs/form-actions and couldn't think of a scenario where this would be a limitation with a non-SSR approach. Do you have an example you could share as I could just be missing something here.

Are there any other features that you wouldn't be able to use on client-side JS (or that you'd rather not) that you could elaborate on?

Slightly related to your comment on an adapter for Tauri (but again for static and not SSR), we're looking at adding zero-config support to SvelteKit's `adapter-static` that will automatically detect a Tauri environment and apply the appropriate settings: https://github.com/sveltejs/kit/pull/9473.

Thanks again for joining the conversation!

**Reply by dummdidumm** [reply]
If you're using SvelteKit's form actions (e.g. having a `+page.server.js` with a `export const actions = ..` in it) then you need a server runtime which can execute that JavaScript. AFAIK that's not something that Tauri offers, which is why you can't use this feature. IMO this is not a drawback of using SvelteKit but rather a general limitation - you can use the other features of SvelteKit just fine. In the context of form interactions it means you'd have to resort to other backend interaction solutions (intercepting the form post request, sending a JSON request to your API and work with the result).

**Reply by Cobular** [reply]
+1 to what @dummdidumm just said - I believe that's correct? On that note, an adapter that moves these form action handlers, or maybe the whole concept of serverside JS, into Tauri commands might be really cool! It might also be too much of a special platform feature to add for just one framework, but with some hacking it might just work since the server files have a very limited set of valid exports.

I do think the `adapter-static` thing would be great, and it would be even better if we could get into the `adapter-auto` list! Iirc, that won't auto switch to static if it's in an unknown environment so we might need explicit support. Regardless, either or both of these would be great!

Regarding other features, there's not a ton else that I can think of right now that's missing. Pretty sure load functions can work on either environment and they're the main other data loading system. Anything that could be done with the sveltekit API capabilities really just shouldn't be done that way in a Tauri environment and should be either a real API or a rust command.

I'm really slammed with work for the rest of the week but I'll dig through some of my projects and see if there's anything else I can see that would be hard to implement this weekend, there might be something I'm missing!

**Reply by ghost** [reply]
@lorenzolewis sorry for barging in into the conversation but I think I have a case where SSG is not an option -- SvelteKit apps with API enpoints. A specific case I've stumbled upon is authenticating users via Google OAuth. I have endpoints that are called from the client side, do the necessary operation (e.g., authenticate the user) and return the result back to the client. Requests like these cannot be sent from the client side.

I might be missing something here so please let me know if I'm wrong about this!

**Reply by FabianLars** [reply]
Nah you're not missing anything. API routes only work in SSR mode and therefore probablyyy will never (officially) work in Tauri. Thanks for the feedback btw!!

**Reply by ghost** [reply]
@FabianLars I see... That's a bummer but thanks for the response anyway! I think I'll re-implement the backend logic in Rust and invoke it from the frontend. Time to learn Rust I guess!

**Comment by austenc** [comment]
Chiming in here about something that doesn't work.

Using SvelteKit and have a store that is using Tauri's `fs` module to persist some data to a json file. I wanted to make this data globally available via a Load Function in the `layout.ts` file, but it doesn't work as you'd expect. Gives a `window is not defined` error. Assuming this is because the SSR render features are turned off?

If I use a load function that doesn't use the tauri `fs` module it works fine. 

Happy to expand or provide code, but it's one limitation of SvelteKit + Tauri I ran into today and wanted to share in case it's helpful. Thank you!

**Reply by probablykasper** [reply]
You'll need to disable SSR. `window is not defined` means the code is running in Node.js, where there's no `window`

**Reply by austenc** [reply]
I have `export const ssr = false` in the `src/routes/layout.ts` file, shouldn't that disable it? 
Normal stores work fine, but creating a custom store that uses things like `exists`, and `writeFile` doesn't seem to. 

The custom store works in components, just not a layout-level Svelte load function for some reason. An example of the modules I tried to use:
```
import { exists, createDir, writeFile, BaseDirectory, readTextFile } from '@tauri-apps/api/fs'
```

If this is something that _should_ work (the load functions, specifically) I am happy to make an example repo so that folks who know more than me can take a look at a reproduction. It's certainly possible I'm doing something wrong!

Thank you!

**Reply by probablykasper** [reply]
You forgot a `+` in the filename, it should be `+layout.ts`

**Comment by Merovex** [comment]
My experience in two words: Thank You! Deep and heartfelt. This scratches an itch I've had for _decades_.

Mind if I share my success story to date?

I was a developer when I was young, and the world was new. I stuck to web development because GUI was a PITA in Perl or other languages, and I've never been one for systems languages. There have been a few things I wanted to GUI (side projects), but Electron was bloated, and with rusty developer skills learning React or Vue was a non-starter. I did not have time to learn Rust in depth. But I knew Tauri + SvelteKit (T+SK) had promise. I tried with T+SK a while back, but it's hard to pour knowledge into a rock.

I write / publish SciFi (side gig), and I enjoy journaling my progress (word counts, mood). My editor keeps the daily wordcount history for me, so I've had scripts that updated my merged journal and created a contribution map (like GHs). Other authors have asked me for my app, but I couldn't get it into their hands.

Enter T+SK, attempt two (electric boogaloo). I used GPT 4 to bridge my knowledge gap. It taught me how to properly configure the Menu, message between the two layers, and debug functions I wrote in Rust. The Dev version ran fine. I ran into problems with the Prod mode (404 error). I went back to the [S+TK Getting Started](https://tauri.app/v1/guides/getting-started/setup/sveltekit/). Well documented, even for a guy like me who hates RTFM. I found my problem (I had `./` instead of `../`). I have fallen in love with SvelteKit, and am falling in love with Rust. Now for me to go through my laundry list.

I could have **benefited from some better code examples**. GPT4 was current enough to get me close enough, so I could get the rest of the way there:

- Menu creation, with Accelerators (I don't think I found this in the documentation, I had to go to source code).
- How to use the application title on the menu (in Windows), still don't know how.
- The message process b/w Tauri + Svelte (how a menu selection can trigger routing to a new page).

<img width="800" alt="Screenshot 2023-04-08 at 12 21 31 PM" src="https://user-images.githubusercontent.com/108851/230732100-ae8ea0ff-3778-4c24-9f79-dabe0ad80bb3.png">

**Comment by bjmc** [comment]
Trying to follow the [quickstart guide here](https://tauri.app/v1/guides/getting-started/setup/sveltekit) and I've completed all the steps (perhaps incorrectly?) but `cargo tauri dev` gives this error:

```
$ cargo tauri dev
     Running BeforeDevCommand (`yarn run dev`)
yarn run v1.22.5
error Couldn't find a package.json file in "/home/bjmc/Sandbox/mytauriproject"
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
       Error The "beforeDevCommand" terminated with a non-zero status code
```

I'm guessing it's because `yarn create svelte` has created the sveltekit project in a subdirectory, rather than in the project root. My folder structure looks like:

```
./svelte/
./src-tauri/
```

I'm sure this is a simple glitch with an easy fix, but this is my first experience with Tauri and it's disappointing that the steps in the docs don't appear to produce a working Hello World app in my case.

**Reply by ghostdevv** [reply]
You'll need to run the tauri init command inside the folder `create-svelte` generates (from this I assume it's your `svelte` folder)

**Reply by FabianLars** [reply]
^this and also both npm install commands should happen in that svelte dir

**Reply by bjmc** [reply]
Thanks @ghostdevv and @FabianLars - that makes perfect sense. I'm not sure how to best concisely capture this in the docs (maybe just an instruction to `cd` into the freshly-created project folder?), but it would help to clarify that the intended result is a root directory containing a Svelte project, and then the `src-tauri/` folder lives inside that.

**Reply by FabianLars** [reply]
I'm not sure either, that's why there's nothing 😂 Some guides actually have a cd instruction but it's so easy to miss it barely makes a difference as far as i can tell.

**Comment by Stijn-B** [comment]
I made a reusable template back when I was looking into this as well https://github.com/auros-one/tauri-sveltekit

Got to the same conclusions as @lorenzolewis' https://github.com/tauri-apps/tauri/discussions/6423#discussioncomment-5352675 except I didn't come in contact with CSP as I use it for a 100% local / offline app.

Happy to collab on a PR if there's anything missing from the tauri-sveltekit template for your usecase

**Comment by fev4** [comment]
Hi there! We're building a PWA with SvelteKit and have been looking into Tauri as possible future step to be able to improve our distribution.

I had a couple of questions that maybe someone in the community can help us with. But first, some context, our app is meant to run offline and heavily depends on service workers and indexeddb (via Dexie.js). The service worker will always start from the app root, which is located in `/app`. We also plan to have a bunch of marketing pages that don't really matter for the potential desktop or mobile apps, like the home page or the `/blog`.

In this sense, the questions we have are:
- Is it possible to select only the `/app` route and everything underneath as the target to be compiled as either desktop and/or mobile apps?
- Can we continue to use our service worker and indexeddb or do we need to handle this differently moving forward if we wanna use Tauri?

Any help or comment is greatly appreciated!

**Reply by probablykasper** [reply]
Determining what files are necessary for `/app` isn't possible for Tauri, so it comes down to SvelteKit:
- If you prerender `/blog`, you could try using [`entries` ](https://kit.svelte.dev/docs/page-options#entries). I can't find any better way to conditionally render routes
- It you want `/app` to be serves as the root, that can be done using [`reroute`](https://kit.svelte.dev/docs/hooks#universal-hooks-reroute)
- There might be a way to have a separate SvelteKit project for `/app` and embed that in your app
- https://github.com/sveltejs/kit/issues/9407 could be helpful if that got implemented

You can use service workers with [tauri-plugin-localhost](https://crates.io/crates/tauri-plugin-localhost), but the ideal way is usually to store data locally instead.
